File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }

        int n = s.length();
        boolean[][] dp = new boolean[n][n];
        int maxLength = 0;
        int start = 0;

        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
            maxLength = 1;
            start = i;
        }

        for (int len = 2; len <= n; len++) {
            for (int i = 0; i < n - len + 1; i++) {
                int j = i + len - 1;

                if (len == 2) {
                    dp[i][j] = (s.charAt(i) == s.charAt(j));
                } else {
                    dp[i][j] = (s.charAt(i) == s.charAt(j)) && dp[i + 1][j - 1];
                }

                if (dp[i][j] && len > maxLength) {
                    maxLength = len;
                    start = i;
                }
            }
        }

        return s.substring(start, start + maxLength);
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.rs
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        if s.is_empty() {
            return String::new();
        }

        let n = s.len();
        let mut dp = vec![vec![false; n]; n];
        let mut max_length = 0;
        let mut start = 0;

        for i in 0..n {
            dp[i][i] = true;
            max_length = 1;
            start = i;
        }

        for len in 2..=n {
            for i in 0..n - len + 1 {
                let j = i + len - 1;

                if len == 2 {
                    dp[i][j] = s.as_bytes()[i] == s.as_bytes()[j];
                } else {
                    dp[i][j] = s.as_bytes()[i] == s.as_bytes()[j] && dp[i + 1][j - 1];
                }

                if dp[i][j] && len > max_length {
                    max_length = len;
                    start = i;
                }
            }
        }

        s[start..start + max_length].to_string()
    }
}
